src/python/pants/core/util_rules/adhoc_process_support.py

```diff
@@ -542,6 +552,57 @@ async def run_adhoc_process(
     return AdhocProcessResult(result, adjusted)
 
 
+# Compute a stable hash value for a `PathMetadata` since the hash value should be stable when used outside the process.
+# (for example, in the cache). The `__hash__` dunder method computes an unstable hash
+# which can and does vary across different process invocations.
+#
+# Ultimately, this may be more of an intellectual correctness point than a necessity since this code will likely never
+# see the same `m.modified` value. It does matter, however, for a single user to see the same
+# behavior across process invocations if pantsd restarts.
+def _path_metadata_to_bytes(m: PathMetadata | None) -> bytes:
+    # Helper to convert a PathMetadataKind to an integer.
+    def kind_to_int(kind: PathMetadataKind) -> int:
+        if kind == PathMetadataKind.FILE:
+            return 0
+        elif kind == PathMetadataKind.DIRECTORY:
+            return 1
+        elif kind == PathMetadataKind.SYMLINK:
+            return 2
+        else:
+            raise TypeError(f"Unable to discriminate instance of `PathMetadataKind`: {repr(kind)}")
+
+    if m is None:
+        return b""
+
+    return struct.pack(
+        "sBQBIsss",
```

Comment:
I think there's a foot-gun with the `s` format specifier that makes this incorrect.

The docs (https://docs.python.org/3/library/struct.html) say (emphasis mine):

> For the 's' format character, the count is interpreted as the **length of the bytes**, not a repeat count like for the other format characters; for example, '10s' means a single 10-byte string mapping to or from a single Python byte string, ... If a count is not given, **it defaults to 1**. For packing, the **string is truncated or padded with null bytes** as appropriate to make it fit.

In particular, this seems to mean the `s` here just take the first byte of each of the inputs, which seems incorrect! The dates will always be `2`, truncating `20xy-...`.

```python
import struct
# reduced example:
struct.pack("s", b"abc")
# b'a'

# for this code:
struct.pack("sBQBIsss", b"path/to/file", 1, 123, 1, 0o755, b"2024-06-21T12:00:00Z", b"2024-06-21T13:00:00Z", b"")
# b'p\x01\x00\x00\x00\x00\x00\x00{\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\xed\x01\x00\x0022\x00'
#   ^                                                                                         ^^
```

(I've put `^` under the strings truncated to their first character.)

It looks like there's no way to get a variable length field using the `struct` module.

Some ideas I thought of:

1. give these over-large fixed widths, e.g. 1000 for paths, 40 for dates:
  ```suggestion
          "1000sBQBI40s40s1000s",
  ```
2. using a different encoding, e.g. 'stable JSON'. There's a conceptually similar function for getting a stable fingerprint of a set of options that could be used as inspiration: https://github.com/pantsbuild/pants/blob/ad47fe4b629d84ec305214c4dff7d7a6077c49ce/src/python/pants/option/options_fingerprinter.py#L27-L33
3. exposing a "stable hash" method from Rust, which can use `#[derive(Hash)]` and an appropriately  chosen hasher. (Not sure how to exclude things like access time.)
4. Something else?

Reply:
Interesting. I missed that part of the pack docs. Stable JSON sounds like the better approach in any event. I'll switch to that.

Reply:
Done.